File: StrukturDataLinkedList/singly_linkedlist.py
class Node:
    def __init__(self, nim, nama):
        self.nim = nim
        self.nama = nama
        self.next = None


class SinglyLinkedList:
    def __init__(self):
        self.head = None
        self.count = 0

    # 1. Insert at beginning
    def insert_beginning(self):
        nim = input("Input NIM  : ")
        nama = input("Input Nama : ")

        new_node = Node(nim, nama)
        new_node.next = self.head
        self.head = new_node
        self.count += 1

    # 2. Insert at given position
    def insert_position(self):
        pos = int(input("Input posisi (1 - {}): ".format(self.count + 1)))

        if pos < 1 or pos > self.count + 1:
            print("Posisi tidak valid")
            return

        nim = input("Input NIM  : ")
        nama = input("Input Nama : ")

        new_node = Node(nim, nama)

        if pos == 1:
            new_node.next = self.head
            self.head = new_node
        else:
            temp = self.head
            for _ in range(pos - 2):
                temp = temp.next
            new_node.next = temp.next
            temp.next = new_node

        self.count += 1

    # 3. Insert at end
    def insert_end(self):
        nim = input("Input NIM  : ")
        nama = input("Input Nama : ")

        new_node = Node(nim, nama)

        if self.head is None:
            self.head = new_node
        else:
            temp = self.head
            while temp.next:
                temp = temp.next
            temp.next = new_node

        self.count += 1

    # 4. Delete from beginning
    def delete_beginning(self):
        if self.head is None:
            print("Data kosong")
            return

        self.head = self.head.next
        self.count -= 1

    # 5. Delete given position
    def delete_position(self):
        pos = int(input("Input posisi (1 - {}): ".format(self.count)))

        if pos < 1 or pos > self.count:
            print("Posisi tidak valid")
            return

        if pos == 1:
            self.head = self.head.next
        else:
            temp = self.head
            for _ in range(pos - 2):
                temp = temp.next
            temp.next = temp.next.next

        self.count -= 1

    # 6. Delete from end
    def delete_end(self):
        if self.head is None:
            print("Data kosong")
            return

        if self.head.next is None:
            self.head = None
        else:
            temp = self.head
            while temp.next.next:
                temp = temp.next
            temp.next = None

        self.count -= 1

    # 7. Delete first occurrence by NIM
    def delete_first_occurrence(self):
        if self.head is None:
            print("Data kosong")
            return

        target = input("Input NIM yang dihapus : ")

        if self.head.nim == target:
            self.head = self.head.next
            self.count -= 1
            return

        temp = self.head
        while temp.next:
            if temp.next.nim == target:
                temp.next = temp.next.next
                self.count -= 1
                return
            temp = temp.next

        print("Data tidak ditemukan")

    # 8. Show data
    def show_data(self):
        if self.head is None:
            print("Data kosong")
            return

        temp = self.head
        while temp:
            print(temp.nim, "-", temp.nama)
            temp = temp.next


# ===== MENU UTAMA =====
ll = SinglyLinkedList()

while True:
    print("\n=== MENU SINGLY LINKED LIST (Data: {}) ===".format(ll.count))
    print("1. Insert at beginning")
    print("2. Insert at given position")
    print("3. Insert at end")
    print("4. Delete from beginning")
    print("5. Delete given position")
    print("6. Delete from end")
    print("7. Delete first occurrence")
    print("8. Show data")
    print("9. Exit")

    pilih = int(input("Pilih menu : "))

    if pilih == 1:
        ll.insert_beginning()
    elif pilih == 2:
        ll.insert_position()
    elif pilih == 3:
        ll.insert_end()
    elif pilih == 4:
        ll.delete_beginning()
    elif pilih == 5:
        ll.delete_position()
    elif pilih == 6:
        ll.delete_end()
    elif pilih == 7:
        ll.delete_first_occurrence()
    elif pilih == 8:
        ll.show_data()
    elif pilih == 9:
        print("Program selesai")
        break
    else:
        print("Menu tidak valid")

File: StrukturDataLinkedList/SinglyLinkedList.java
import java.util.Scanner;

class Node {
    String nim, nama;
    Node next;

    Node(String nim, String nama) {
        this.nim = nim;
        this.nama = nama;
        this.next = null;
    }
}

public class SinglyLinkedList {
    Node head;
    int count = 0;
    Scanner sc = new Scanner(System.in);

    // 1. Insert at beginning
    void insertBeginning() {
        System.out.print("Input NIM  : ");
        String nim = sc.nextLine();
        System.out.print("Input Nama : ");
        String nama = sc.nextLine();

        Node newNode = new Node(nim, nama);
        newNode.next = head;
        head = newNode;
        count++;
    }

    // 2. Insert at given position
    void insertPosition() {
        System.out.print("Input posisi (1 - " + (count + 1) + ") : ");
        int pos = sc.nextInt();
        sc.nextLine();

        if (pos < 1 || pos > count + 1) {
            System.out.println("Posisi tidak valid");
            return;
        }

        System.out.print("Input NIM  : ");
        String nim = sc.nextLine();
        System.out.print("Input Nama : ");
        String nama = sc.nextLine();

        Node newNode = new Node(nim, nama);

        if (pos == 1) {
            newNode.next = head;
            head = newNode;
        } else {
            Node temp = head;
            for (int i = 1; i < pos - 1; i++) {
                temp = temp.next;
            }
            newNode.next = temp.next;
            temp.next = newNode;
        }
        count++;
    }

    // 3. Insert at end
    void insertEnd() {
        System.out.print("Input NIM  : ");
        String nim = sc.nextLine();
        System.out.print("Input Nama : ");
        String nama = sc.nextLine();

        Node newNode = new Node(nim, nama);

        if (head == null) {
            head = newNode;
        } else {
            Node temp = head;
            while (temp.next != null)
                temp = temp.next;
            temp.next = newNode;
        }
        count++;
    }

    // 4. Delete from beginning
    void deleteBeginning() {
        if (head == null) {
            System.out.println("Data kosong");
            return;
        }
        head = head.next;
        count--;
    }

    // 5. Delete given position
    void deletePosition() {
        System.out.print("Input posisi (1 - " + count + ") : ");
        int pos = sc.nextInt();
        sc.nextLine();

        if (pos < 1 || pos > count) {
            System.out.println("Posisi tidak valid");
            return;
        }

        if (pos == 1) {
            head = head.next;
        } else {
            Node temp = head;
            for (int i = 1; i < pos - 1; i++) {
                temp = temp.next;
            }
            temp.next = temp.next.next;
        }
        count--;
    }

    // 6. Delete from end
    void deleteEnd() {
        if (head == null) {
            System.out.println("Data kosong");
            return;
        }

        if (head.next == null) {
            head = null;
        } else {
            Node temp = head;
            while (temp.next.next != null)
                temp = temp.next;
            temp.next = null;
        }
        count--;
    }

    // 7. Delete first occurrence
    void deleteFirstOccurrence() {
        if (head == null) {
            System.out.println("Data kosong");
            return;
        }

        System.out.print("Input NIM yang dihapus : ");
        String target = sc.nextLine();

        if (head.nim.equals(target)) {
            head = head.next;
            count--;
            return;
        }

        Node temp = head;
        while (temp.next != null) {
            if (temp.next.nim.equals(target)) {
                temp.next = temp.next.next;
                count--;
                return;
            }
            temp = temp.next;
        }

        System.out.println("Data tidak ditemukan");
    }

    // 8. Show data
    void showData() {
        if (head == null) {
            System.out.println("Data kosong");
            return;
        }

        Node temp = head;
        while (temp != null) {
            System.out.println(temp.nim + " - " + temp.nama);
            temp = temp.next;
        }
    }

    public static void main(String[] args) {
        SinglyLinkedList sll = new SinglyLinkedList();
        Scanner sc = new Scanner(System.in);

        while (true) {
            System.out.println("\n=== MENU SINGLY LINKED LIST (Data: " + sll.count + ") ===");
            System.out.println("1. Insert at beginning");
            System.out.println("2. Insert at given position");
            System.out.println("3. Insert at end");
            System.out.println("4. Delete from beginning");
            System.out.println("5. Delete given position");
            System.out.println("6. Delete from end");
            System.out.println("7. Delete first occurrence");
            System.out.println("8. Show data");
            System.out.println("9. Exit");
            System.out.print("Pilih menu : ");

            int pilih = sc.nextInt();
            sc.nextLine();

            switch (pilih) {
                case 1 -> sll.insertBeginning();
                case 2 -> sll.insertPosition();
                case 3 -> sll.insertEnd();
                case 4 -> sll.deleteBeginning();
                case 5 -> sll.deletePosition();
                case 6 -> sll.deleteEnd();
                case 7 -> sll.deleteFirstOccurrence();
                case 8 -> sll.showData();
                case 9 -> {
                    System.out.println("Program selesai");
                    return;
                }
                default -> System.out.println("Menu tidak valid");
            }
        }
    }
}
